python/mlx/nn/layers/recurrent.py

```diff
@@ -0,0 +1,542 @@
+# Copyright © 2023 Apple Inc.
+
+import math
+
+import mlx.core as mx
+from mlx.nn.layers import Dropout
+from mlx.nn.layers.base import Module
+
+
+class RecurrentCellBase(Module):
+    r"""Base class for Recurrent cells. Helper for shared
+    attributes and weights initialization.
+
+        Args:
+        input_size (int): Dimension of the input :math:`x`
+        hidden_size (int): Dimension of the hidden state(s)
+        bias (bool): Whether to use biases or not
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        bias: bool,
+    ):
+        super().__init__()
+        self.input_size = input_size
+        self.hidden_size = hidden_size
+        self.bias = bias
+
+    def _init_weights(self):
+        scale = 1.0 / math.sqrt(self.hidden_size)
+        Wx = mx.random.uniform(
+            low=-scale, high=scale, shape=(self.input_size, self.hidden_size)
+        )
+        Wh = self.Wxh = mx.random.uniform(
+            low=-scale, high=scale, shape=(self.hidden_size, self.hidden_size)
+        )
+
+        if self.bias:
+            bx = mx.random.uniform(low=-scale, high=scale, shape=(self.hidden_size,))
+            bh = mx.random.uniform(low=-scale, high=scale, shape=(self.hidden_size,))
+        else:
+            bx = bh = None
+
+        return Wx, Wh, bx, bh
+
+
+class RNNCell(RecurrentCellBase):
+    r"""Apply one step of Elman recurrent unit with tanh or relu
+    non-linearity.
+
+    Concretely:
+
+    .. math::
+
+        h' = \text{tanh} (W_{xh}x + b_{xh} + W_{hh}h + b_{hh})
+
+    The input :math:`x` has shape ``ND` or ``D`` and the hidden
+    state :math:`h` has shape ``NH`` or ``H``, where:
+        - ``N`` is the batch dimension
+        - ``D`` is the input's features dimension
+        - ``H`` is the hidden dimension
+
+    If `nonlinearity` is `relu`, then ReLU is used instead of tanh.
+
+    Args:
+        input_size (int): Dimension of the input :math:`x`
+        hidden_size (int): Dimension of the hidden state :math:`h`
+        nonlinearity (str): Non-linearity to use. Valid values: `tanh`
+        or `relu`. Default: `tanh`.
+        bias (bool): Whether to use biases or not. Default: `True`.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        hidden_size: int,
+        nonlinearity: str = "tanh",
```

Comment:
how about make this a callable instead and set it to a reasonable default (tanh?) ? I'm not certain I like it more but it is a lot more flexible

Reply:
Sure, that works as well